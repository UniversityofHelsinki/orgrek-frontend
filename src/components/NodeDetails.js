/* eslint-disable max-lines-per-function */
/* eslint-disable max-lines */
import React, { useEffect, useState } from 'react';
import { connect } from 'react-redux';
import NodeDetailsTable from './NodeDetailsTable';
import { filterAttributeDuplicates, datesOverlap } from '../actions/utilAction';
import { switchHistory, switchComing, updateAttributes } from '../actions/nodeViewAction';

import {
    fetchNodeParents,
    fetchNodeChildren
} from '../actions/hierarchyAction';
import {
    fetchNode,
    fetchNodeAttributes,
    fetchNodeFavorableFullNames,
    fetchNodeFullNames,
    fetchNodePredecessors,
    fetchNodeSuccessors
} from '../actions/nodeAction';
import { useTranslation } from 'react-i18next';
import { codeAttributes } from '../constants/variables';
import EditButtons from './EditButtons';
import { isAdmin } from '../actions/userAction';
import moment from 'moment';
import NewUpperUnit from './NewUpperUnit';

// eslint-disable-next-line complexity
const NodeDetails = (props) => {
    const { t, i18n } = useTranslation();
    const lang = i18n.language;
    const [attributeData, setAttributeData] = useState(false);
    const [modified, setModified] = useState({});
    const [modifiedParents, setModifiedParents] = useState({});

    const uniqueIdAttribute = props.node
        ? { 'key': 'unique_id', 'value': props.node.uniqueId, startDate: null, endDate: null }
        : { 'key': 'unique_id', 'value': t('no_value'), startDate: null, endDate: null };

    const sortOtherAttributes = elems => {
        const sortedList = elems.sort((a, b) => {
            if (!a.startDate && !a.endDate) {
                return -1;
            }
            if (!b.startDate && !b.endDate) {
                return 1;
            }
            if (b.endDate && a.startDate) {
                return new Date(b.endDate) - new Date(a.startDate);
            }
            return new Date(a.startDate) - new Date(b.startDate);
        });
        return sortedList;
    };

    const byCodesAndDates = (a,b) => {
        if (a.key === b.key) {
            if (!a.startDate && !a.endDate) {
                return -1;
            }
            if (!b.startDate && !b.endDate) {
                return 1;
            }
            if (b.endDate && a.startDate) {
                return new Date(b.endDate) - new Date(a.startDate);
            }
            return new Date(a.startDate) - new Date(b.startDate);
        }
        const aOrder = codeAttributes.findIndex(key => key === a.key);
        const bOrder = codeAttributes.findIndex(key => key === b.key);
        if (aOrder < bOrder) {
            return -1;
        } else if (aOrder > bOrder) {
            return 1;
        }
        return 0;
    };

    const sortNameAttributesByDate = (elems, order) => {
        let result = [];
        for (let property in order) {
            const filteredBatch = elems.filter(e => {
                return e.key === property;
            });
            filteredBatch.sort((a,b) => {
                return (!(a.endDate || b.endDate) && 0) || (!a.endDate && -1) || (!b.endDate && 1) || new Date(b.endDate) - new Date(a.endDate);
            });
            result.push(filteredBatch);
        }
        return result.flat();
    };

    const orderNameAttributesByLanguage = (elems) => {
        const order = { name_fi : 0, name_sv : 1, name_en : 2, default: 3 };
        elems.sort((a,b) => order[a.key] - order[b.key]);
        return sortNameAttributesByDate(elems, order);
    };

    const selectData = () => {
        if (props.showHistory && props.showComing && props.nodeAttributesHistory && props.nodeAttributes) {
            setAttributeData(filterAttributeDuplicates(props.nodeAttributesHistory, props.nodeAttributesFuture));
        } else if (props.showHistory && props.nodeAttributesHistory) {
            setAttributeData(props.nodeAttributesHistory);
        } else if (props.showComing && props.nodeAttributesFuture) {
            setAttributeData(props.nodeAttributesFuture);
        } else {
            setAttributeData(props.nodeAttributes);
        }
    };
    const nameInfoData = attributeData ? attributeData.filter(elem => /^name_(fi|sv|en)$/.test(elem.key)) : false;
    const nameInfoDataOrderedByLanguage = nameInfoData ? orderNameAttributesByLanguage(nameInfoData) : false;
    const typeAttributeData = attributeData ? attributeData.filter(elem => elem.key === 'type') : false;
    const codeAttributesData = attributeData
        ? [uniqueIdAttribute, ...attributeData.filter(a => codeAttributes.includes(a.key))]
            .sort(byCodesAndDates)
            .filter(a => !nameInfoData.includes(a) && !typeAttributeData.includes(a))
        : false;
    const otherAttributesData = attributeData ? attributeData.filter(elem => {
        return !nameInfoData.includes(elem) && !typeAttributeData.includes(elem) && !codeAttributesData.includes(elem);
    }) : false;
    const sortedOtherAttributesData = otherAttributesData ? sortOtherAttributes(otherAttributesData) : false;
    let validityData = props.node ? [props.node] : false;

    const isCurrent = datesOverlap(
        props.node?.startDate && new Date(Date.parse(props.node.startDate)),
        props.node?.endDate && new Date(Date.parse(props.node.endDate)),
        props.selectedDay
    );
    const isPast = !isCurrent && props.node?.endDate && new Date(Date.parse(props.node.endDate)).getTime() <= props.selectedDay.getTime();
    const isFuture = !isCurrent && props.node?.startDate && new Date(Date.parse(props.node.startDate)).getTime() >= props.selectedDay.getTime();

    useEffect(() => {
        if (props.node) {
            validityData = props.node ? [props.node] : false;
            console.log(props.node);
            console.log(validityData);
            const startDate = Date.parse(props.node.startDate) || undefined;
            const endDate = Date.parse(props.node.endDate) || undefined;
            if (datesOverlap(startDate && new Date(startDate), endDate && new Date(endDate), props.selectedDay)) {
                props.fetchNodeDetails(props.node, props.selectedDay, props.showHistory, props.showComing, props.selectedHierarchy);
            } else if (endDate && new Date(endDate).getTime() <= props.selectedDay.getTime()) {
                props.fetchNodeDetails(props.node, new Date(endDate), props.showHistory, props.showComing, props.selectedHierarchy);
            } else if (startDate && new Date(startDate).getTime() >= props.selectedDay.getTime()) {
                props.fetchNodeDetails(props.node, new Date(startDate), props.showHistory, props.showComing, props.selectedHierarchy);
            }
        }
    }, [props.node, props.showComing, props.showHistory, props.selectedHierarchy]);

    React.useLayoutEffect(() => {
        selectData();
    }, [props.node, props.nodeAttributes, props.nodeAttributesFuture, props.nodeAttributesHistory]);

    const pastFutureFilter = (data) => {
        if (isCurrent || props.showComing || props.showHistory) {
            return data;
        }
        return [];
    };

    const onValueChange = (event, elem) => {
        if (modified[elem.id]) {//element has already been modified at least once, because its found in modified array
            const target = { ...modified[elem.id], [event.target.name]: event.target.value , validity: false, 'err': '' };//makes copy of modified[elem.id] and updates its value with event.target.value
            setModified({ ...modified, [elem.id]: target });//updates row
        } else {//This is the first time this element is modified so its not found in modified array
            const target = { ...elem, [event.target.name]: event.target.value, validity: false, 'err': '' };//creates a new elem object based on elem object and updates its value with event.target.value
            setModified({ ...modified, [elem.id]: target, 'err': '' });//adds this new object in modified map
        }
    };


    const onDateChange = (dateChanged) => {
        let elem = dateChanged.elem;
        let date = moment(dateChanged.date).utcOffset(0).format('YYYY-MM-DDTHH:mm:ss.sss+00:00');
        if (date === 'Invalid date') {
            date = null;
        }
        let name = dateChanged.whichDate; //startDate or endDate

        if (modified[elem.id]) {//element has already been modified at least once, because its found in modified array
            const target = { ...modified[elem.id], [name]: date, validity: dateChanged.validity, 'err': '' };//makes copy of modified[elem.id] and updates its value with date
            setModified({ ...modified, [elem.id]: target });//updates row
        } else {//This is the first time this element is modified so its not found in modified array
            const target = { ...elem, [name]: date, validity: dateChanged.validity, 'err': '' };//creates a new elem object based on elem object and updates its value with date
            setModified({ ...modified, [elem.id]: target });//adds this new object in modified map
    };

    const onParentDateChange = (dateChanged) => {
        let elem = dateChanged.elem;
        let date = moment(dateChanged.date).utcOffset(0).format('YYYY-MM-DDTHH:mm:ss.sss+00:00');
        if (date === 'Invalid date') {
            date = null;
        }
        let name = dateChanged.whichDate;
        let hierarchyIndex = elem.hierarchies.findIndex((obj => obj.hierarchy === dateChanged.hierarchy));
        if (name === 'startDate') {
            elem.hierarchies[hierarchyIndex].startDate = date;
        }
        if (name === 'endDate') {
            elem.hierarchies[hierarchyIndex].endDate = date;
        }
        if (modifiedParents[elem.id]) {
            setModifiedParents({ ...modifiedParents, [elem.id]: elem });
        } else {
            setModifiedParents({ ...modifiedParents, [elem.id]: elem });
        }
    };

    return (
        <div>
            {props.nodeAttributes &&
                <>
                    {isAdmin(props.user) ? <EditButtons modifiedParents={modifiedParents} setModified={setModified} node={props.node} selectedDay={props.selectedDay} selectedHierarchy={props.selectedHierarchy} modified={modified} /> : null }
                    <div className="right-side">
                        <div>
                            <h3>{props.favorableNames[lang === 'ia' && 'fi' || lang]?.[0]?.name}</h3>
                        </div>
                        <NodeDetailsTable
                            selectedDay={props.selectedDay}
                            type='key-value'
                            heading='valid_dates'
                            tableLabels={[]}
                            contentData={validityData ? [...validityData.map((elem => {
                                if (modified[elem.id]) {
                                    return modified[elem.id];
                                }
                                return elem;
                            }))] : validityData }
                            hasValidity={true}
                            onValueChange={onValueChange}
                            onDateChange={onDateChange}
                            fullname={false}
                        />
                        <NodeDetailsTable
                            selectedDay={props.selectedDay}
                            type='key-value'
                            heading='name_info'
                            tableLabels={['text_language_header', 'name']}
                            contentData={nameInfoDataOrderedByLanguage ? [...nameInfoDataOrderedByLanguage.map((elem => {
                                if (modified[elem.id]) {
                                    return modified[elem.id];
                                }
                                return elem;
                            }))] : nameInfoDataOrderedByLanguage }
                            hasValidity={true}
                            dataFilter={pastFutureFilter}
                            onValueChange={onValueChange}
                            onDateChange={onDateChange}
                            fullname={false}
                        />
                        <NodeDetailsTable
                            selectedDay={props.selectedDay}
                            type='key-value'
                            heading='display_name_info'
                            tableLabels={['text_language_header', 'name']}
                            contentData={[...(props.displayNames.fi || []), ...(props.displayNames.sv || []), ...(props.displayNames.en || [])].filter(n => n).map(dn => ({ ...dn, key: `name_${dn.language.toLowerCase()}`, value: dn.name }))}
                            hasValidity={true}
                            dataFilter={pastFutureFilter}
                            fullname={true}
                            onValueChange={onValueChange}
                            onDateChange={onDateChange}
                        />
                        <NodeDetailsTable
                            selectedDay={props.selectedDay}
                            type='key-value'
                            heading='codes'
                            tableLabels={['code_namespace', 'value']}
                            contentData={codeAttributesData ? [...codeAttributesData.map((elem => {
                                if (modified[elem.id]) {
                                    return modified[elem.id];
                                }
                                return elem;
                            }))]: codeAttributesData }
                            hasValidity={true}
                            dataFilter={data => (isPast || isFuture) && !(props.showHistory || props.showComing) ? data.filter(attr => attr.key === 'unique_id') : data}
                            onValueChange={onValueChange}
                            onDateChange={onDateChange}
                            fullname={false}
                        />
                        <NodeDetailsTable
                            selectedDay={props.selectedDay}
                            type='key-value'
                            heading='unit_type'
                            tableLabels={[]}
                            contentData={typeAttributeData ? [...typeAttributeData.map((elem => {
                                if (modified[elem.id]) {
                                    return modified[elem.id];
                                }
                                return elem;
                            }))]: typeAttributeData }
                            hasValidity={true}
                            dataFilter={pastFutureFilter}
                            onValueChange={onValueChange}
                            onDateChange={onDateChange}
                            fullname={false}
                        />
                        <NodeDetailsTable
                            selectedDay={props.selectedDay}
                            type='node-hierarchy'
                            heading='upper_units'
                            tableLabels={['unit', 'hierarchies', 'hierarchy_valid']}
                            contentData={props.parents[lang === 'ia' && 'fi' || lang]}
                            hasValidity={false}
                            dataFilter={pastFutureFilter}
                            onValueChange={onValueChange}
                            onDateChange={onParentDateChange}
                            fullname={false}
                        />
                        <NewUpperUnit />
                        <NodeDetailsTable
                            selectedDay={props.selectedDay}
                            type='node-hierarchy'
                            heading='subunits'
                            tableLabels={['unit', 'hierarchies', 'hierarchy_valid']}
                            contentData={props.children[lang === 'ia' && 'fi' || lang]}
                            hasValidity={false}
                            dataFilter={pastFutureFilter}
                            onValueChange={onValueChange}
                            onDateChange={onDateChange}
                            fullname={false}
                        />
                        <NodeDetailsTable
                            selectedDay={props.selectedDay}
                            type='name-validity'
                            heading='predecessors'
                            tableLabels={['name', 'valid_dates', 'predecessor_edge_valid']}
                            contentData={props.predecessors[lang === 'ia' && 'fi' || lang]}
                            hasValidity={false}
                            onValueChange={onValueChange}
                            onDateChange={onDateChange}
                            fullname={false}
                        />
                        <NodeDetailsTable
                            selectedDay={props.selectedDay}
                            type='name-validity'
                            heading='successors'
                            tableLabels={['name', 'valid_dates', 'successor_edge_valid']}
                            contentData={props.successors[lang === 'ia' && 'fi' || lang]}
                            hasValidity={false}
                            onValueChange={onValueChange}
                            onDateChange={onDateChange}
                            fullname={false}
                        />
                        <NodeDetailsTable
                            selectedDay={props.selectedDay}
                            type='key-value'
                            heading='other_attributes'
                            tableLabels={['attribute', 'value']}
                            contentData={sortedOtherAttributesData ? [...sortedOtherAttributesData.map((elem => {
                                if (modified[elem.id]) {
                                    return modified[elem.id];
                                }
                                return elem;
                            }))] : sortedOtherAttributesData }
                            hasValidity={true}
                            dataFilter={pastFutureFilter}
                            onValueChange={onValueChange}
                            onDateChange={onDateChange}
                            fullname={false}
                        />
                    </div>
                </>
            }
        </div>
    );
};

const mapStateToProps = state => ({
    node : state.nrd.node,
    nodeAttributes : state.nrd.nodeAttributes,
    nodeAttributesFuture: state.nrd.nodeAttributesFuture,
    nodeAttributesHistory: state.nrd.nodeAttributesHistory,
    parents : state.hr.parents,
    parentsHistory: state.hr.parentsHistory,
    parentsFuture: state.hr.parentsFuture,
    children : state.hr.children,
    childrenHistory: state.hr.childrenHistory,
    childrenFuture: state.hr.childrenFuture,
    predecessors : state.nrd.nodePredecessors,
    successors : state.nrd.nodeSuccessors,
    selectedDay: state.dr.selectedDay,
    showHistory: state.nvrd.showHistory,
    showComing: state.nvrd.showComing,
    displayNames: state.nrd.nodeDisplayNames,
    favorableNames: state.nrd.nodeFavorableNames,
    selectedHierarchy: state.tree.selectedHierarchy,
    user : state.ur.user
});

const mapDispatchToProps = dispatch => ({
    onSwitchHistory: (input) => {
        dispatch(switchHistory(input));
    },
    onSwitchComing: (input) => {
        dispatch(switchComing(input));
    },
    fetchNodeDetails: (node, selectedDay, showHistory, showComing, selectedHierarchy) => {
        dispatch(fetchNodePredecessors(node.uniqueId, selectedDay));
        dispatch(fetchNodeSuccessors(node.uniqueId, selectedDay));
        dispatch(fetchNodeFavorableFullNames(node.uniqueId, selectedDay));
        if (!(showHistory || showComing)) {
            dispatch(fetchNodeAttributes(node.uniqueId, selectedDay, selectedHierarchy));
            dispatch(fetchNodeParents(node.uniqueId, selectedDay));
            dispatch(fetchNodeChildren(node.uniqueId, selectedDay));
            dispatch(fetchNodeFullNames(node.uniqueId, selectedDay));
        }
    },
    fetchNode: (uniqueId) => dispatch(fetchNode(uniqueId))
});

export default connect(mapStateToProps, mapDispatchToProps)(NodeDetails);
