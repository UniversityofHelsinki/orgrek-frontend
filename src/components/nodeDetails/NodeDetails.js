import React from 'react';
import NodeValiditySection from './NodeValiditySection';
import NameSection from './NameSection';
import DisplayNameSection from './DisplayNameSection';
import CodeAttributesSection from './CodeAttributesSection';
import UnitTypeSection from './UnitTypeSection';
import ParentsSection from './ParentsSection';
import ChildrenSection from './ChildrenSection';
import PredecessorsSection from './PredecessorsSection';
import SuccessorsSection from './SuccessorsSection';
import OtherAttributesSection from './OtherAttributesSection';
import Typography from '@mui/material/Typography';
import IfAdmin from '../../auth/IfAdmin';
import useFavorableName from '../../hooks/useFavorableName';
import useFetchNodeDetails from '../../hooks/useFetchNodeDetails';
import Box from '@mui/material/Box';
import useFetchNode from '../../hooks/useFetchNode';
import { useNodeId } from '../../hooks/useNodeId';

const NodeDetails = () => {
  const nodeId = useNodeId();
  const title = useFavorableName();

  // These legacy fetch hooks can be removed after everything has been migrated
  // to use RTK Query
  useFetchNode();
  useFetchNodeDetails();

  if (!nodeId) {
    return null;
  }

  return (
    <Box
      key={nodeId}
      component="main"
      sx={{
        pl: 4,
        pr: 4,
        pb: 8,
        flex: 1,
      }}
    >
      <Typography variant="h1" component="h2" mt={5} mb={5}>
        {title}
      </Typography>
      <NodeValiditySection />
      <NameSection />
      <DisplayNameSection />
      <CodeAttributesSection />
      <UnitTypeSection />
      <ParentsSection />
      <ChildrenSection />
      <PredecessorsSection />
      <SuccessorsSection />
      <OtherAttributesSection />
    </Box>
  );
};

export default NodeDetails;
