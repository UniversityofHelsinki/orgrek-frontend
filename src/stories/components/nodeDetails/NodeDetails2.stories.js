import { expect } from '@storybook/jest';
import React from 'react';
import NodeDetails from '../../../components/nodeDetails/NodeDetails2';
import {
  createAdmin,
  createNodeState,
  createReader,
  hierarchyFilters,
  mockGetAttributeKeys,
  mockGetAttributes,
  mockGetChildren,
  mockGetCodeAttributes,
  mockGetFavorableFullNames,
  mockGetFullNames,
  mockGetNameAttributes,
  mockGetNode,
  mockGetParents,
  mockGetPredecessors,
  mockGetSectionTypeAttributes,
  mockGetSuccessors,
  mockGetTypeAttributes,
  mockGetValidHierarchyFilters,
  mockPutNameAttributes,
  mockPutTypeAttributes,
  mockSaveCodeAttributes,
  withMockStore,
} from '../../../mockStore';
import { waitFor, within } from '@storybook/testing-library';

// Use a fixed date to ensure that tests always have a consistent result
const now = new Date('2023-03-22T14:28:00+0200');

const nodeId = '1';
const selectedHierarchy = 'talous';

export default {
  component: NodeDetails,
  parameters: {
    layout: 'fullscreen',
    systemTime: now,
    reactRouter: {
      searchParams: {
        uid: nodeId,
      },
    },
  },
};

export const AdminRole = {
  parameters: {
    msw: {
      handlers: [
        mockGetNode(nodeId, {
          id: '4820',
          name: 'TIKE, Tietotekniikkaratkaisut (TIRA)',
          startDate: null,
          endDate: null,
          timestamp: '2022-01-18T15:39:54.603+00:00',
          uniqueId: nodeId,
        }),
        mockGetSectionTypeAttributes('names', [
          {
            id: 14,
            section: 'names',
            attr: 'name_fi',
          },
          {
            id: 15,
            section: 'names',
            attr: 'name_sv',
          },
          {
            id: 16,
            section: 'names',
            attr: 'name_en',
          },
        ]),
        mockGetSectionTypeAttributes('types', [
          {
            id: 17,
            section: 'types',
            attr: 'type',
          },
        ]),
        mockGetNameAttributes(nodeId, [
          {
            id: 4899999926,
            nodeId: '4820',
            key: 'name_fi',
            value: 'Tietotekniikkaratkaisut 1',
            startDate: '1999-12-31',
            endDate: null,
          },
          {
            id: 48111122,
            nodeId: '4820',
            key: 'name_sv',
            value: 'Datatekniklösningar',
            startDate: '1999-12-31',
            endDate: null,
          },
          {
            id: 480021,
            nodeId: '4820',
            key: 'name_en',
            value: 'IT Solutions 1',
            startDate: '1999-12-31',
            endDate: null,
          },
        ]),
        mockPutNameAttributes(nodeId),
        mockGetCodeAttributes(nodeId, [
          {
            id: 4824,
            nodeId: '4820',
            key: 'talous_tunnus',
            value: 'H9073',
            startDate: null,
            endDate: null,
          },
          {
            id: 4823,
            nodeId: '4820',
            key: 'lyhenne',
            value: 'TIRA',
            startDate: null,
            endDate: null,
          },
          {
            id: 14528,
            nodeId: '4820',
            key: 'emo_lyhenne',
            value: 'TIKE',
            startDate: null,
            endDate: null,
          },
        ]),
        mockSaveCodeAttributes(nodeId),
        mockGetTypeAttributes(nodeId, [
          {
            id: 1,
            nodeId: '1',
            key: 'type',
            value: 'tulosyksikko',
            startDate: null,
            endDate: null,
          },
        ]),
        mockPutTypeAttributes(nodeId),
        mockGetAttributeKeys(
          {
            selectedHierarchies: selectedHierarchy,
            sections: ['codes'],
          },
          ['emo_lyhenne', 'lyhenne', 'talous_tunnus']
        ),
        mockGetAttributeKeys(
          {
            selectedHierarchies: selectedHierarchy,
            sections: ['types'],
          },
          ['type']
        ),
        mockGetValidHierarchyFilters(now, hierarchyFilters),
        mockGetFullNames(nodeId, now, {
          sv: [
            {
              nodeId: '4820',
              language: 'sv',
              name: 'TIKE, Datatekniklösningar (TIRA)',
              startDate: null,
              endDate: null,
            },
          ],
          fi: [
            {
              nodeId: '4820',
              language: 'fi',
              name: 'TIKE, Tietotekniikkaratkaisut (TIRA)',
              startDate: null,
              endDate: null,
            },
          ],
          en: [
            {
              nodeId: '4820',
              language: 'en',
              name: 'TIKE, IT Solutions (TIRA)',
              startDate: null,
              endDate: null,
            },
          ],
        }),
        mockGetFavorableFullNames(nodeId, now, {
          sv: [
            {
              nodeId: '4820',
              language: 'sv',
              name: 'TIKE, Datatekniklösningar (TIRA)',
              startDate: null,
              endDate: null,
            },
          ],
          fi: [
            {
              nodeId: '4820',
              language: 'fi',
              name: 'TIKE, Tietotekniikkaratkaisut (TIRA)',
              startDate: null,
              endDate: null,
            },
          ],
          en: [
            {
              nodeId: '4820',
              language: 'en',
              name: 'TIKE, IT Solutions (TIRA)',
              startDate: null,
              endDate: null,
            },
          ],
        }),
        mockGetParents(nodeId, now, { fi: [], sv: [], en: [] }),
        mockGetChildren(nodeId, now, { fi: [], sv: [], en: [] }),
        mockGetPredecessors(nodeId, now, {
          sv: [],
          fi: [],
          en: [],
        }),
        mockGetSuccessors(nodeId, now, {
          sv: [],
          fi: [],
          en: [],
        }),
        // Remove after all sections have been migrated to RTK Query
        mockGetAttributes(nodeId, now, selectedHierarchy, [
          {
            id: 27788,
            nodeId: '4820',
            key: 'publicity',
            value: 'julkinen',
            startDate: null,
            endDate: null,
          },
        ]),
      ],
    },
  },
  decorators: [
    withMockStore({
      dr: { selectedDay: now },
      nrd: createNodeState({
        node: { uniqueId: nodeId },
      }),
      tree: { selectedHierarchy },
      ur: { user: createAdmin() },
    }),
  ],
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);

    await waitFor(async () => {
      expect(canvas.getAllByTestId('attributesTable')).toHaveLength(5);
    });
  },
};

export const ReaderRole = {
  ...AdminRole,
  decorators: [
    withMockStore({
      dr: { selectedDay: now },
      nrd: createNodeState({
        node: { uniqueId: nodeId },
      }),
      tree: { selectedHierarchy },
      ur: { user: createReader() },
    }),
  ],
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);

    await waitFor(async () => {
      expect(canvas.getAllByTestId('attributesTable')).toHaveLength(5);
    });
  },
};
