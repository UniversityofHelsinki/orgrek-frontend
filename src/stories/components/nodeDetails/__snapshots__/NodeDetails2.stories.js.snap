// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stories/components/nodeDetails/NodeDetails2 AdminRole smoke-test 1`] = `
Object {
  "checked": undefined,
  "children": undefined,
  "name": "stories-components-nodedetails-nodedetails2--admin-role",
  "pressed": undefined,
  "role": "WebArea",
  "value": undefined,
}
`;

exports[`stories/components/nodeDetails/NodeDetails2 ReaderRole smoke-test 1`] = `
Object {
  "checked": undefined,
  "children": undefined,
  "name": "stories-components-nodedetails-nodedetails2--reader-role",
  "pressed": undefined,
  "role": "WebArea",
  "value": undefined,
}
`;
